clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java

```diff
@@ -353,134 +360,145 @@ public void testMetadataFetchOnStaleMetadata() throws Exception {
-            producer.send(extendedRecord, null);
+            producer.send(extendedRecord);
             fail("Expected KafkaException to be raised");
         } catch (KafkaException e) {
             // expected
         }
-        PowerMock.verify(metadata);
-
-        // Expect exactly two fetches if topic metadata is available but outdated for the given record
-        PowerMock.reset(metadata);
-        EasyMock.expect(metadata.fetch()).andReturn(initialCluster).once();
-        EasyMock.expect(metadata.fetch()).andReturn(extendedCluster).once();
-        EasyMock.expect(metadata.fetch()).andThrow(new IllegalStateException("Unexpected call to metadata.fetch()")).anyTimes();
-        PowerMock.replay(metadata);
-        producer.send(extendedRecord, null);
-        PowerMock.verify(metadata);
+        verify(metadata, times(5)).requestUpdate();
+        verify(metadata, times(5)).awaitUpdate(anyInt(), anyLong());
+        verify(metadata, times(8)).fetch();
+
+        // One request update if metadata is available but outdated for the given record
+        producer.send(extendedRecord);
+        verify(metadata, times(6)).requestUpdate();
+        verify(metadata, times(6)).awaitUpdate(anyInt(), anyLong());
+        verify(metadata, times(10)).fetch();
+
+        producer.close(0, TimeUnit.MILLISECONDS);
     }
 
     @Test
-    public void testTopicRefreshInMetadata() throws Exception {
+    public void testTopicRefreshInMetadata() throws InterruptedException {
         Properties props = new Properties();
         props.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
         props.setProperty(ProducerConfig.MAX_BLOCK_MS_CONFIG, "600000");
-        KafkaProducer<String, String> producer = new KafkaProducer<>(props, new StringSerializer(), new StringSerializer());
+        ProducerConfig config = new ProducerConfig(ProducerConfig.addSerializerToConfig(props, new StringSerializer(),
+            new StringSerializer()));
         long refreshBackoffMs = 500L;
         long metadataExpireMs = 60000L;
         final Metadata metadata = new Metadata(refreshBackoffMs, metadataExpireMs, true,
                 true, new ClusterResourceListeners());
         final Time time = new MockTime();
-        MemberModifier.field(KafkaProducer.class, "metadata").set(producer, metadata);
-        MemberModifier.field(KafkaProducer.class, "time").set(producer, time);
         final String topic = "topic";
-
-        Thread t = new Thread(() -> {
-            long startTimeMs = System.currentTimeMillis();
-            for (int i = 0; i < 10; i++) {
-                while (!metadata.updateRequested() && System.currentTimeMillis() - startTimeMs < 1000)
-                    Thread.yield();
-                metadata.update(Cluster.empty(), Collections.singleton(topic), time.milliseconds());
-                time.sleep(60 * 1000L);
+        try (KafkaProducer<String, String> producer = new KafkaProducer<>(config, null, null, metadata,
```

Comment:
Would it be more realistic and consistent with old code to do `new KafkaProducer<>(config, new StringSerializer(), new StringSerializer(), metadata, null, null, time))` here? The new code at line 488 seems to be consistent with the old code.

Reply:
It's a good question. The problem is that we have to set the serializers in `ProducerConfig`. Once we do that, we can just pass `null` for both of these to make it obvious that they have been set already. Otherwise, you can end up with inconsistent serializers by updating them in one place and forgetting to update them elsewhere.

This is ugly though and I wasn't too happy with it, so I changed the producer a bit to simplify this. Let me know what you think.

Reply:
Previously the test code used constructor `KafkaProducer(Map<String, Object> configs, Serializer<K> keySerializer, Serializer<V> valueSerializer)`, which translates to `KafkaProducer(new ProducerConfig(ProducerConfig.addSerializerToConfig(configs, keySerializer, valueSerializer)),keySerializer, valueSerializer, null, null, null, Time.SYSTEM)` where keySerializer and valueSerializer are not null. I was thinking that the goal this patch is to mimic the same behavior.

Now I look through the `KafkaProducer` code again, I realized there is constructor `KafkaProducer(final Map<String, Object> configs)` which translates to `KafkaProducer(new ProducerConfig(configs), null, null, null, null, null, Time.SYSTEM)`. It is OK for the test code to switch to this constructor after explicitly using `addSerializerToConfig(...)`.

It looks like a lot of test failed after the most recent commit. Maybe it is simpler to just revert the recent change and focus on the migration to mockito in this PR.

Reply:
Btw, I left line 488 as it is because there's what I'd call a bug in `ProducerConfig.addSerializerToConfig` where it doesn't use the actual instance, just the class name. However, it's a public method so not sure if people are relying on it. This will go away in the PR that has the commit I just reverted so I think we can merge as it is.